be/src/storage/segment_sync_executor.cpp

```diff
@@ -0,0 +1,285 @@
+            LOG(WARNING) << "Failed to send rpc to " << debug_string() << " err=" << _st;
+            return _st;
+        }
+        _st = _closure->result.status();
+        if (!_st.ok()) {
+            LOG(WARNING) << "Failed to send rpc to " << debug_string() << " err=" << _st;
+            return _st;
+        }
+
+        for (size_t i = 0; i < _closure->result.tablet_vec_size(); ++i) {
+            sync_tablet_infos->emplace_back(std::make_unique<PTabletInfo>());
+            sync_tablet_infos->back()->Swap(_closure->result.mutable_tablet_vec(i));
+        }
+    }
+
+    return Status::OK();
+}
+
+void SyncChannel::cancel() {
+    if (!_init().ok()) {
+        return;
+    }
+
+    // cancel rpc request, accelerate the release of related resources
+    // Cancel an already-cancelled call_id has no effect. 
+    _closure->cancel();
+
+    PTabletWriterCancelRequest request;
+    request.set_allocated_id(const_cast<starrocks::PUniqueId*>(&_opt->load_id));
+    request.set_sender_id(0);
+    request.set_tablet_id(_opt->tablet_id);
+    request.set_txn_id(_opt->txn_id);
+    request.set_index_id(_opt->index_id);
+
+    auto closure = new ReusableClosure<PTabletWriterCancelResult>();
+
+    closure->ref();
+    closure->cntl.set_timeout_ms(_opt->timeout_ms);
+    _stub->tablet_writer_cancel(&closure->cntl, &request, &closure->result, closure);
+    request.release_id();
+
+    VLOG(1) << "Cancel request " << debug_string();
+}
+
+SyncToken::SyncToken(std::unique_ptr<ThreadPoolToken> sync_pool_token, const DeltaWriterOptions* opt)
+        : _sync_token(std::move(sync_pool_token)), _status(), _opt(opt), _fs(std::move(new_fs_posix())) {
+    // first replica is primary replica, skip it
+    for (size_t i = 1; i < opt->replicas.size(); ++i) {
+        _sync_channels.emplace_back(std::move(std::make_unique<SyncChannel>(
+                opt, opt->replicas[i].host(), opt->replicas[i].port(), opt->replicas[i].node_id())));
+    }
+    // default quorom policy
+    _max_fail_replica_num = opt->replicas.size() - (opt->replicas.size() / 2 + 1);
+}
+
+Status SyncToken::submit(std::unique_ptr<SegmentPB> segment, bool eos) {
+    RETURN_IF_ERROR(status());
+    auto task = std::make_shared<SegmentSyncTask>(this, std::move(segment), eos);
+    return _sync_token->submit(std::move(task));
+}
+
+void SyncToken::cancel() {
+    for (auto& channel : _sync_channels) {
+        channel->cancel();
+    }
+    _sync_token->shutdown();
+}
+
+Status SyncToken::wait() {
+    _sync_token->wait();
+    std::lock_guard l(_status_lock);
+    return _status;
+}
+
+std::string SyncToken::debug_string() {
+    return fmt::format("[SyncToken tablet_id: {}, txn_id: {}]", _opt->tablet_id, _opt->txn_id);
+}
+
+void SyncToken::_sync_segment(std::unique_ptr<SegmentPB> segment, bool eos) {
+    // If previous sync has failed, return directly
+    if (!status().ok()) return;
+
+    // 1. read segment from local storage
```

Comment:
Why not sync_segment before the main segment is flushed? From the memory will be fast.

Reply:
there are two reason now
1. segment_writer write through buffer io directly, not write to a buffer, then flush it. If we were to do this, we would need to change the write logic a lot and use more memory. At the same time, the recently written segments are generally in the page cache, so the reading overhead is not too large.
2. When the replication is delayed (it is easy to happen through the network), the segment buffer will be backlogged, which will take up a lot of memory.